File: library/src/test/java/me/proxer/library/api/user/UserHistoryEndpointTest.java
package me.proxer.library.api.user;

import me.proxer.library.ProxerTest;
import me.proxer.library.api.ProxerException;
import me.proxer.library.entity.user.UserHistoryEntry;
import me.proxer.library.enums.Category;
import me.proxer.library.enums.MediaLanguage;
import me.proxer.library.enums.Medium;
import okhttp3.mockwebserver.MockResponse;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class UserHistoryEndpointTest extends ProxerTest {

    @Test
    void testDefault() throws ProxerException, IOException {
        server.enqueue(new MockResponse().setBody(fromResource("user_history.json")));

        final List<UserHistoryEntry> result = api.user().history("test", "supersecret")
                .build()
                .execute();

        assertThat(result).first().isEqualTo(buildFirstTestEntry());
        assertThat(result).last().isEqualTo(buildLastTestEntry());
    }

    @Test
    void testPath() throws ProxerException, IOException, InterruptedException {
        server.enqueue(new MockResponse().setBody(fromResource("user_history.json")));

        api.user().history("test", "supersecret")
                .page(1)
                .limit(12)
                .includeHentai(true)
                .build()
                .execute();

        assertThat(server.takeRequest().getPath()).isEqualTo("/api/v1/user/history?uid=test&username=supersecret"
                + "&p=1&limit=12&isH=true");
    }

    private UserHistoryEntry buildFirstTestEntry() {
        return new UserHistoryEntry("457484352", "16209", "Kono Subarashii Sekai ni Shukufuku wo! 2",
                MediaLanguage.ENGLISH_SUB, Medium.ANIMESERIES, Category.ANIME, 2);
    }

    private UserHistoryEntry buildLastTestEntry() {
        return new UserHistoryEntry("456582967", "3088", "Girls of the Wild's",
                MediaLanguage.ENGLISH, Medium.MANGASERIES, Category.MANGA, 182);
    }
}

File: library/src/test/kotlin/me/proxer/library/api/user/UserHistoryEndpointTest.kt
package me.proxer.library.api.user

import me.proxer.library.ProxerTest
import me.proxer.library.entity.user.UserHistoryEntry
import me.proxer.library.enums.Category
import me.proxer.library.enums.MediaLanguage
import me.proxer.library.enums.Medium
import me.proxer.library.runRequest
import org.amshove.kluent.shouldBeEqualTo
import org.junit.jupiter.api.Test

/**
 * @author Ruben Gees
 */
class UserHistoryEndpointTest : ProxerTest() {

    private val firstExpectedEntry = UserHistoryEntry(
        id = "457484352",
        entryId = "16209",
        name = "Kono Subarashii Sekai ni Shukufuku wo! 2",
        language = MediaLanguage.ENGLISH_SUB,
        medium = Medium.ANIMESERIES,
        category = Category.ANIME,
        episode = 2
    )

    private val secondExpectedEntry = UserHistoryEntry(
        id = "456582967",
        entryId = "3088",
        name = "Girls of the Wild's",
        language = MediaLanguage.ENGLISH,
        medium = Medium.MANGASERIES,
        category = Category.MANGA,
        episode = 182
    )

    @Test
    fun testDefault() {
        val (result, _) = server.runRequest("user_history.json") {
            api.user.history("test", "supersecret")
                .build()
                .safeExecute()
        }

        result.first() shouldBeEqualTo firstExpectedEntry
        result.last() shouldBeEqualTo secondExpectedEntry
    }

    @Test
    fun testPath() {
        val (_, request) = server.runRequest("user_history.json") {
            api.user.history("test", "supersecret")
                .page(1)
                .limit(12)
                .includeHentai(true)
                .build()
                .execute()
        }

        request.path shouldBeEqualTo "/api/v1/user/history?uid=test&username=supersecret&p=1&limit=12&isH=true"
    }
}
